src/openms/source/ANALYSIS/OPENSWATH/PeakIntegrator.cpp

```diff
@@ -375,4 +375,706 @@ namespace OpenMS
+        if (iter_idx < 10000 && iter_idx / info_iter_threshold >= 10) info_iter_threshold *= 10;
+      }
+
+      // If the cost function doesn't change enough, the gradient descent algorithm is terminated
+      // This is decided by computing the standard deviation between a selection of the last few Es
+      if (iter_idx % 100 == 0) // TODO: maybe use a different value than 1000, otherwise a minimum of 10k iterations is needed
+      {
+        last_few_Es[last_few_Es_idx++ % last_few_Es_dim] = current_E;
+        const double mean = std::accumulate(last_few_Es.begin(), last_few_Es.end(), 0.0) / last_few_Es_dim;
+        double squared_diffs {0.0};
+        for (const double current_E : last_few_Es)
+        {
+          squared_diffs += std::pow(current_E - mean, 2.0);
+        }
+        const double Es_std_dev = std::sqrt(squared_diffs / static_cast<double>(last_few_Es_dim));
+        if (Es_std_dev < Es_std_dev_min)
+        {
+          if (print_debug_ == 1)
+          {
+            std::cout << std::endl << "[" << iter_idx << "] The cost function is not changing enough, anymore. Breaking.";
+            std::cout << std::endl << "[" << iter_idx << "] [mean=" << mean << "] [Es_std_dev=" << Es_std_dev << "]" << std::endl;
+          }
+          break;
+        }
+      }
+
+      // Simultaneous update of all parameters for gradient descent
+      iRpropPlus(prev_diff_E_h, diff_E_h, lr_h, term_h, h, current_E, previous_E);
+      iRpropPlus(prev_diff_E_mu, diff_E_mu, lr_mu, term_mu, mu, current_E, previous_E);
+      iRpropPlus(prev_diff_E_sigma, diff_E_sigma, lr_sigma, term_sigma, sigma, current_E, previous_E);
+      iRpropPlus(prev_diff_E_tau, diff_E_tau, lr_tau, term_tau, tau, current_E, previous_E);
+
+      // Apply the parameters' constraints
+      h = std::max(h_lower_boundary, h);
+      if (mu < mu_left_boundary || mu_right_boundary < mu)
+      {
+        mu = mu < mu_left_boundary ? mu_left_boundary : mu_right_boundary;
+      }
+      sigma = std::min(std::max(1e-4, sigma), 20.0); // `sigma` constraints // TODO: find non-magic values for min and max sigmas. maybe remove std::min (the upper boundary)
+      tau = std::min(std::max(sigma, tau), sigma * 15.0); // `tau` constraints // TODO: is this a good constraint?
+
+      // Saving values to be used at the next iteration
+      prev_diff_E_h = diff_E_h;
+      prev_diff_E_mu = diff_E_mu;
+      prev_diff_E_sigma = diff_E_sigma;
+      prev_diff_E_tau = diff_E_tau;
+      previous_E = current_E;
+    }
+    if (print_debug_ == 1)
+    {
+      std::cout << std::endl << "[" << best_iter << "] RESULT: best_E=" << best_E << std::endl;
+      std::cout << "[" << best_iter << "] GEOGEBRA: Execute[{\"h = " << best_h << "\", \"mu = " << best_mu << "\",\"sigma = " << best_sigma << "\", \"tau = " << best_tau << "\"}]" << std::endl;
+    }
+    // Return the number of iterations that occurred.
+    // Note: it is not the index of the best iteration,
```

Comment:
Not sure what you mean here.

Reply:
I'll try to improve the wording.
The algorithm has a maximum number of iterations hardcoded. This limit is rarely reached, and instead a lower number of iterations are run. The method returns such number.